cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -79,38 +84,24 @@ predicate isNonConst(DataFlow::Node node, boolean isIndirect) {
       )
     )
     or
-    exists(Parameter p | p = e.(VariableAccess).getTarget() |
-      p.getFunction().getName() = "main" and p.getType() instanceof PointerType
-    )
-    or
-    e instanceof CrementOperation
-    or
-    e instanceof AddressOfExpr
-    or
-    e instanceof ReferenceToExpr
-    or
-    e instanceof AssignPointerAddExpr
-    or
-    e instanceof AssignPointerSubExpr
-    or
-    e instanceof PointerArithmeticOperation
-    or
-    e instanceof FieldAccess
-    or
-    e instanceof PointerDereferenceExpr
-    or
-    e instanceof AddressOfExpr
-    or
-    e instanceof ExprCall
-    or
-    e instanceof NewArrayExpr
-    or
     exists(Variable v | v = e.(VariableAccess).getTarget() |
       v.getType().(ArrayType).getBaseType() instanceof CharType and
       exists(AssignExpr ae |
         ae.getLValue().(ArrayExpr).getArrayBase().(VariableAccess).getTarget() = v
       )
     )
+    or
+    exists(UncalledFunction f, Parameter p | f.getAParameter() = p |
+      p = e.(VariableAccess).getTarget()
+    )
+    or
+    node instanceof FlowSource
+    or
+    node instanceof DataFlow::DefinitionByReferenceNode and
+    not exists(FormattingFunctionCall fc | node.asDefiningArgument() = fc.getOutputArgument(_)) and
+    not exists(Call c |
+      c.getAnArgument() = node.asDefiningArgument() and c.getTarget().hasDefinition()
+    )
```

Comment:
So (it would be nice if we had had comments on these cases) we have:
- return values from library implemented functions (that are not `gettext`...)
- a case for flow through arrays
- accesses of parameters in functions that are never called (e.g. `main`)
- flow sources
- outputs from out parameters of library implemented functions (that are not `sprintf`...)
- (below) outputs from out parameters (indirect contents only)

I guess the overarching principle is that these are places where a (non-constant) value can enter from outside of the code we track data flow through - whether that be from outside the program itself or just a return value from a library function.  Does this sound about right?

Reply:
I can add comments, and I think I need to expand this but I need your help for that. 

First, what this is intended to do is to say if you call a function with no definition, I will basically have to assume the string you get back is anything. The second `not exist` covers this case. 

Secondly, the first `not exist` is where I need help. In effect, I want to have as a barrier (or prevent any input from being identified) at functions (any function) where we have defined some sort of dataflow through it. In otherwords, let those functions with defined dataflows, just use their dataflows, and do not attempt to consider that an input source. Do you have any suggestions for this? My current approach is to ignore all formatting function calls, but we will instantly hit another issue (as we move more in this direction) where a strcpy is considered an input, and other variants (similar cases) I predict will pop up. Can I say something like `not exists(Function f | f.getAnArgument() = [node.asIndirectionExpr(), node.asExpr()] and f.hasDefinedDataFlow())`?

Reply:
I have a solution for this. It will be in my next push here, but going down the rabbit hole of changing how non-const is found, I eventually find myself with a solution that will regress DCA a bunch and I don't think there is any incremental way around it.

Reply:
Sounds like this might be best done as a follow-up PR then.  There's already a lot going on in this pull request, and a change that regresses results may complicate getting it merged.